spark/src/test/java/org/apache/iceberg/actions/TestExpireSnapshotsAction.java

```diff
@@ -0,0 +1,780 @@
+
+public abstract class TestExpireSnapshotsAction extends SparkTestBase {
+
+  private static final HadoopTables TABLES = new HadoopTables(new Configuration());
+  private static final Schema SCHEMA = new Schema(
+      optional(1, "c1", Types.IntegerType.get()),
+      optional(2, "c2", Types.StringType.get()),
+      optional(3, "c3", Types.StringType.get())
+  );
+
+  private static final PartitionSpec SPEC = PartitionSpec.builderFor(SCHEMA).identity("c1").build();
+
+  private static final List<ThreeColumnRecord> RECORDS = Lists.newArrayList(new ThreeColumnRecord(1, "AAAA", "AAAA"));
+
+  static final DataFile FILE_A = DataFiles.builder(SPEC)
+      .withPath("/path/to/data-a.parquet")
+      .withFileSizeInBytes(10)
+      .withPartitionPath("c1=0") // easy way to set partition data for now
+      .withRecordCount(1)
+      .build();
+  static final DataFile FILE_B = DataFiles.builder(SPEC)
+      .withPath("/path/to/data-b.parquet")
+      .withFileSizeInBytes(10)
+      .withPartitionPath("c1=1") // easy way to set partition data for now
+      .withRecordCount(1)
+      .build();
+  static final DataFile FILE_C = DataFiles.builder(SPEC)
+      .withPath("/path/to/data-c.parquet")
+      .withFileSizeInBytes(10)
+      .withPartitionPath("c1=2") // easy way to set partition data for now
+      .withRecordCount(1)
+      .build();
+  static final DataFile FILE_D = DataFiles.builder(SPEC)
+      .withPath("/path/to/data-d.parquet")
+      .withFileSizeInBytes(10)
+      .withPartitionPath("c1=3") // easy way to set partition data for now
+      .withRecordCount(1)
+      .build();
+
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
+
+  private File tableDir;
+  private String tableLocation;
+  private Table table;
+
+  @Before
+  public void setupTableLocation() throws Exception {
+    this.tableDir = temp.newFolder();
+    this.tableLocation = tableDir.toURI().toString();
+    this.table = TABLES.create(SCHEMA, SPEC, Maps.newHashMap(), tableLocation);
+  }
+
+  private Dataset<Row> buildDF(List<ThreeColumnRecord> records) {
+    return spark.createDataFrame(records, ThreeColumnRecord.class).coalesce(1);
+  }
+
+  private void writeDF(Dataset<Row> df, String mode) {
+    df.select("c1", "c2", "c3")
+        .write()
+        .format("iceberg")
+        .mode(mode)
+        .save(tableLocation);
+  }
+
+  private void checkExpirationResults(Long expectedDatafiles, Long expectedManifestsDeleted,
+      Long expectedManifestListsDeleted, ExpireSnapshotsActionResult results) {
+
+    Assert.assertEquals("Incorrect number of manifest files deleted",
+        expectedManifestsDeleted, results.getManifestFilesDeleted());
+    Assert.assertEquals("Incorrect number of datafiles deleted",
+        expectedDatafiles, results.getDataFilesDeleted());
+    Assert.assertEquals("Incorrect number of manifest lists deleted",
+        expectedManifestListsDeleted, results.getManifestListsDeleted());
+  }
+
```

Comment:
The only tests not directly ported were those doing file cleanup (except for the parallel delete one) since we already had new tests for those functions.